src/resolveElements.js

```diff
@@ -0,0 +1,104 @@
+import isPromise from 'is-promise';
+import React from 'react';
+import warning from 'warning';
+
+import HttpError from './HttpError';
+
+const UNRESOLVED = {};
+
+function checkResolved(value) {
+  if (!isPromise(value)) {
+    return value;
+  }
+
+  return Promise.race([value, UNRESOLVED]);
+}
+
+function isResolved(value) {
+  return value !== UNRESOLVED;
+}
+
+function createElements(match, Components, matchData) {
+  const { routes, routeParams: matchRouteParams } = match;
+
+  return routes.map((route, i) => {
+    const Component = Components[i];
+    const data = matchData[i];
+    const routeParams = matchRouteParams[i];
+
+    const routeMatch = {
+      ...match,
+      route,
+      routeParams: routeParams[i],
+    };
+
+    if (route.render) {
+      // Perhaps undefined here would be more correct for "not ready", but
+      // Relay uses null in RelayReadyStateRenderer, so let's follow that
+      // example.
+      return route.render({
+        Component: isResolved(Component) ? Component : null,
+        props: isResolved(data) ? { ...routeMatch, data } : null,
+        match: routeMatch,
+      });
+    }
+
+    if (!isResolved(Component) || !isResolved(data)) {
+      // Can't render.
+      return null;
+    }
+
+    if (!Component) {
+      warning(
+        data === undefined,
+        `Route ${i} has data, but no render method or component.`,
```

Comment:
this is going to be the worlds most unhelpful warning if `i` is just an index :P